### PR TITLE
Constructor consistency update

### LINKED ISSUE #2416: [RFC] How to apply configuration
* `Uart::new()` should be infallible, only applying some invalid config should be invalid
* ~~Remove the independent `change_x` mutator methods in favour of `apply_config(&mut self, config: &Config)`~~ https://github.com/esp-rs/esp-hal/pull/2449

### PR DESCRIPTION
**Pull request by bugadani**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] I have added necessary changes to user code to the [Migration Guide](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/MIGRATING-0.21.md).
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md) - Well, I changed the guide 🤔🫠

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

#### Description

Closes #2416

### CODE REVIEW

**Review by bjoernQ — APPROVED** [review]
LGTM

File: documentation/API-GUIDELINES.md [hunk 1]
```diff
@@ -22,14 +22,18 @@ The following paragraphs contain additional recommendations.
   - The peripheral instance type must default to a type that supports any of the peripheral instances.
   - The author must to use `crate::any_peripheral` to define the "any" peripheral instance type.
   - The driver must implement a `new` constructor that automatically converts the peripheral instance into the any type, and a `new_typed` that preserves the peripheral type.
+- If a driver is configurable, configuration options should be implemented as a `Config` struct in the same module where the driver is located.
+  - The driver's constructor should take the config struct by value, and it should return `Result<Self, ConfigError>`.
+  - The driver should implement `fn apply_config(&mut self, config: &Config) -> Result<(), ConfigError>`.
+  - In case the driver's configuration is infallible (all possible combinations of options are supported by the hardware), the `ConfigError` should be implemented as an empty `enum`.
```
[the remark this answers is not in the document]
Comment (bugadani) [reply documentation/API-GUIDELINES.md:29]: I was hoping that the name implies this, but added a line.

**Review by jessebraham — APPROVED** [review]
LGTM, thanks!

**Review by MabezDev — APPROVED** [review]
Thanks!

**Pull request merged**